File: backend/audio_analyzer.py
"""
Audio Analyzer for Waveform Visualization
Analyzes audio files using FFT to generate real-time waveform visualization data.
"""

from PySide6.QtCore import QObject, Signal, Slot, QTimer
from concurrent.futures import ThreadPoolExecutor
import os

from backend.logging_config import get_logger
logger = get_logger(__name__)

# Try to import numpy for FFT computation
try:
    import numpy as np
    NUMPY_AVAILABLE = True
except ImportError:
    NUMPY_AVAILABLE = False
    logger.warning("NumPy not available - waveform visualization will be disabled")

# Try to import av for audio decoding
try:
    import av
    AV_AVAILABLE = True
except ImportError:
    AV_AVAILABLE = False
    logger.warning("PyAV not available - waveform visualization will be disabled")


class AudioAnalyzer(QObject):
    """
    Analyzes audio files to generate FFT visualization data.
    Pre-computes FFT for entire track, then provides data based on playback position.
    """

    # Signal emitted with FFT levels as a list of integers (0-8)
    fftDataChanged = Signal(list)

    # Signal emitted when analysis is complete
    analysisComplete = Signal()

    # Signal emitted when analysis starts
    analysisStarted = Signal()

    def __init__(self, parent=None):
        super().__init__(parent)

        # Quality tier → bar count mapping
        self._quality_bars = {"Low": 16, "Medium": 32, "High": 48, "Extreme": 64, "Insane": 96}

        # FFT data storage - list of lists, each inner list is FFT levels for a time chunk
        self._fft_data = []
        self._current_file = ""
        self._num_bars = 32
        self._chunk_duration = 0.1  # 100ms chunks (10 chunks per second)

        # Current levels emitted to QML (no animation, just the looked-up data)
        self._current_levels = [0] * self._num_bars

        # Track if we're actively playing
        self._is_active = False

        # Thread pool for off-thread analysis
        self._executor = ThreadPoolExecutor(max_workers=1)
        self._analyzing = False
        self._pending_future = None

        # Poll timer — checks if the worker thread finished, keeps Qt work on the main thread
        self._poll_timer = QTimer(self)
        self._poll_timer.setInterval(100)
        self._poll_timer.timeout.connect(self._check_analysis_done)

    @Slot(str)
    def analyze_file(self, file_path: str):
        """
        Analyze an audio file and pre-compute FFT visualization data.
        Called when a new track starts playing. Runs off the main thread.
        """
        logger.info(f"analyze_file called with: {file_path}")

        if not AV_AVAILABLE or not NUMPY_AVAILABLE:
            logger.warning("PyAV or NumPy not available, skipping analysis")
            return

        if not file_path or not os.path.exists(file_path):
            logger.warning(f"Invalid file path for analysis: {file_path}")
            return

        # Skip if already analyzed
        if file_path == self._current_file and self._fft_data:
            logger.info(f"File already analyzed: {file_path}")
            return

        # Skip if already analyzing
        if self._analyzing:
            logger.info("Analysis already in progress, skipping")
            return

        self._current_file = file_path
        self._analyzing = True

        self.analysisStarted.emit()
        logger.info(f"Starting audio analysis for: {file_path}")

        # Submit work to the thread pool and poll for completion from the main thread
        self._pending_future = self._executor.submit(self._analyze_audio, file_path)
        self._poll_timer.start()

    def _check_analysis_done(self):
        """Poll callback — runs on the main thread via QTimer."""
        if self._pending_future is None or not self._pending_future.done():
            return

        self._poll_timer.stop()
        future = self._pending_future
        self._pending_future = None
        self._analyzing = False

        try:
            result = future.result()
            if result is not None:
                self._fft_data = result
                self.analysisComplete.emit()
                logger.info(f"Analysis complete: {len(self._fft_data)} chunks generated")
            else:
                self._fft_data = []
        except Exception as e:
            logger.error(f"Error in audio analysis thread: {e}", exc_info=True)
            self._fft_data = []

    def _analyze_audio(self, file_path: str):
        """
        Decode audio and compute FFT for visualization.
        Runs on a worker thread — must not touch Qt objects.
        Returns the computed FFT data list, or None on failure.
        """
        try:
            container = av.open(file_path)
            audio_stream = next((s for s in container.streams if s.type == 'audio'), None)

            if not audio_stream:
                logger.debug("No audio stream found")
                return None

            # Decode all audio frames
            samples = []
            sample_rate = audio_stream.rate or 44100

            for frame in container.decode(audio_stream):
                frame_data = frame.to_ndarray()

                # If stereo, convert to mono by averaging channels
                if len(frame_data.shape) > 1:
                    frame_data = frame_data.mean(axis=0)

                samples.append(frame_data)

            container.close()

            if not samples:
                logger.debug("No audio samples decoded")
                return None

            # Concatenate all samples
            all_samples = np.concatenate(samples).astype(np.float32)

            # Normalize samples
            max_val = np.max(np.abs(all_samples))
            if max_val > 0:
                all_samples = all_samples / max_val

            # Downsample to ~8kHz for faster processing
            downsample_factor = max(1, sample_rate // 8000)
            all_samples = all_samples[::downsample_factor]
            effective_rate = sample_rate // downsample_factor

            # Compute FFT for time chunks
            chunk_size = int(effective_rate * self._chunk_duration)
            num_chunks = len(all_samples) // chunk_size

            if num_chunks == 0:
                logger.debug("Audio too short for analysis")
                return None

            # Pre-compute window function once
            window = np.hanning(chunk_size)

            # First pass: collect raw FFT magnitudes
            raw_fft_data = []
            for i in range(num_chunks):
                chunk = all_samples[i * chunk_size:(i + 1) * chunk_size]

                # Apply window function to reduce spectral leakage
                windowed = chunk * window

                # Compute FFT
                fft = np.abs(np.fft.rfft(windowed))
                fft = fft[1:len(fft)//2]  # Skip DC, use lower half

                if len(fft) == 0:
                    raw_fft_data.append([0.0] * self._num_bars)
                    continue

                # Split into frequency bands with logarithmic spacing
                log_indices = np.logspace(0, np.log10(len(fft)), self._num_bars + 1, dtype=int)
                log_indices = np.clip(log_indices, 0, len(fft))

                levels = []
                for j in range(self._num_bars):
                    start_idx = log_indices[j]
                    end_idx = log_indices[j + 1]
                    if end_idx > start_idx:
                        band = fft[start_idx:end_idx]
                        levels.append(np.mean(band))
                    else:
                        levels.append(0.0)

                raw_fft_data.append(levels)

            # Find global max for normalization (use 95th percentile to avoid outliers)
            all_values = [v for chunk in raw_fft_data for v in chunk if v > 0]
            if all_values:
                global_max = np.percentile(all_values, 95)
            else:
                global_max = 1.0

            # Second pass: normalize and map to 0-8 range
            fft_data = []
            for levels in raw_fft_data:
                normalized = []
                for level in levels:
                    if global_max > 0:
                        norm = level / global_max
                        norm = min(1.0, norm) ** 0.6
                        normalized.append(int(norm * 8))
                    else:
                        normalized.append(0)
                fft_data.append(normalized)

            return fft_data

        except Exception as e:
            logger.error(f"FFT analysis error: {e}")
            return None

    @Slot(float)
    def update_position(self, position_seconds: float):
        """
        Look up pre-computed FFT levels for the current playback position
        and emit them directly to QML. No intermediate animation — QML
        handles any smoothing it needs.
        """
        if not self._fft_data or not self._is_active:
            return

        chunk_index = int(position_seconds * 10)
        chunk_index = max(0, min(chunk_index, len(self._fft_data) - 1))

        new_levels = self._fft_data[chunk_index]

        # Only emit if levels actually changed
        if new_levels != self._current_levels:
            self._current_levels = new_levels
            self.fftDataChanged.emit(new_levels)

    @Slot(bool)
    def set_active(self, active: bool):
        """Enable or disable the visualizer."""
        logger.info(f"set_active called with: {active}, has FFT data: {len(self._fft_data)} chunks")
        self._is_active = active
        if not active:
            # Emit zeros to clear the visualizer
            self._current_levels = [0] * self._num_bars
            self.fftDataChanged.emit(self._current_levels)

    @Slot(result=list)
    def get_current_levels(self) -> list:
        """Get current FFT levels for QML."""
        return self._current_levels

    @Slot(result=int)
    def get_num_bars(self) -> int:
        """Get number of frequency bars."""
        return self._num_bars

    @Slot(result=bool)
    def is_analyzed(self) -> bool:
        """Check if current file has been analyzed."""
        return len(self._fft_data) > 0

    @Slot()
    def clear(self):
        """Clear analysis data."""
        self._fft_data = []
        self._current_file = ""
        self._current_levels = [0] * self._num_bars

    @Slot(str)
    def set_quality(self, quality: str):
        """Change quality tier — adjusts bar count and clears cached data."""
        new_bars = self._quality_bars.get(quality, 32)
        if new_bars == self._num_bars:
            return
        logger.info(f"Visualizer quality changed to {quality}: {new_bars} bars")
        self._num_bars = new_bars
        self._current_levels = [0] * self._num_bars
        # Clear cached analysis so next track (or re-analysis) uses new bar count
        self._fft_data = []
        self._current_file = ""

File: src/managers/audioanalyzer.cpp
#include "audioanalyzer.h"

#include <QUrl>
#include <QFileInfo>
#include <QProcess>
#include <QtConcurrent>
#include <QLoggingCategory>

#include <algorithm>
#include <numeric>
#include <cstring>

Q_LOGGING_CATEGORY(lcAudioAnalyzer, "octave.audioanalyzer")

// ════════════════════════════════════════════════════════════════
// Construction / destruction
// ════════════════════════════════════════════════════════════════

AudioAnalyzer::AudioAnalyzer(QObject *parent)
    : QObject(parent)
{
    m_qualityBars[QStringLiteral("Low")]     = 16;
    m_qualityBars[QStringLiteral("Medium")]  = 32;
    m_qualityBars[QStringLiteral("High")]    = 48;
    m_qualityBars[QStringLiteral("Extreme")] = 64;
    m_qualityBars[QStringLiteral("Insane")]  = 96;

    // Initialise current levels to zeros
    m_currentLevels.reserve(m_numBars);
    for (int i = 0; i < m_numBars; ++i)
        m_currentLevels.append(0);

    connect(&m_watcher, &QFutureWatcher<AnalysisResult>::finished,
            this, &AudioAnalyzer::onAnalysisDone);
}

AudioAnalyzer::~AudioAnalyzer()
{
    m_watcher.cancel();
    m_watcher.waitForFinished();
}

// ════════════════════════════════════════════════════════════════
// Minimal in-place radix-2 Cooley-Tukey FFT
// ════════════════════════════════════════════════════════════════

void AudioAnalyzer::fft_radix2(std::vector<std::complex<float>> &x)
{
    const int N = static_cast<int>(x.size());
    if (N <= 1)
        return;

    // Bit-reversal permutation
    for (int i = 1, j = 0; i < N; ++i) {
        int bit = N >> 1;
        while (j & bit) {
            j ^= bit;
            bit >>= 1;
        }
        j ^= bit;
        if (i < j)
            std::swap(x[i], x[j]);
    }

    // Butterfly stages
    for (int len = 2; len <= N; len <<= 1) {
        const float angle = -2.0f * static_cast<float>(M_PI) / static_cast<float>(len);
        const std::complex<float> wlen(std::cos(angle), std::sin(angle));
        for (int i = 0; i < N; i += len) {
            std::complex<float> w(1.0f, 0.0f);
            for (int j = 0; j < len / 2; ++j) {
                std::complex<float> u = x[i + j];
                std::complex<float> v = x[i + j + len / 2] * w;
                x[i + j]             = u + v;
                x[i + j + len / 2]   = u - v;
                w *= wlen;
            }
        }
    }
}

// ════════════════════════════════════════════════════════════════
// Hann window
// ════════════════════════════════════════════════════════════════

std::vector<float> AudioAnalyzer::hannWindow(int N)
{
    std::vector<float> w(N);
    for (int i = 0; i < N; ++i)
        w[i] = 0.5f * (1.0f - std::cos(2.0f * static_cast<float>(M_PI) * static_cast<float>(i) / static_cast<float>(N)));
    return w;
}

// ════════════════════════════════════════════════════════════════
// Next power of two (for zero-padding to radix-2 length)
// ════════════════════════════════════════════════════════════════

static int nextPow2(int n)
{
    int p = 1;
    while (p < n)
        p <<= 1;
    return p;
}

// ════════════════════════════════════════════════════════════════
// Worker: decode audio file and compute per-chunk FFT levels
// Runs on a QtConcurrent thread — must NOT touch Qt GUI objects.
//
// Uses ffmpeg CLI to decode audio to raw PCM float data, matching
// the Python version's use of PyAV.  QAudioDecoder was previously
// used here but fails silently on worker threads due to event-loop
// and thread-affinity issues with the GStreamer/FFmpeg backends.
// ════════════════════════════════════════════════════════════════

AudioAnalyzer::AnalysisResult AudioAnalyzer::analyzeAudio(
    const QString &filePath, int numBars, double chunkDuration)
{
    AnalysisResult result;

    // ── Decode via ffmpeg CLI → raw 32-bit float, 8 kHz, mono ────
    QProcess proc;
    proc.setProgram(QStringLiteral("ffmpeg"));
    proc.setArguments({
        QStringLiteral("-i"), filePath,
        QStringLiteral("-f"), QStringLiteral("f32le"),   // raw 32-bit float LE
        QStringLiteral("-acodec"), QStringLiteral("pcm_f32le"),
        QStringLiteral("-ar"), QStringLiteral("8000"),   // 8 kHz
        QStringLiteral("-ac"), QStringLiteral("1"),      // mono
        QStringLiteral("-v"), QStringLiteral("quiet"),
        QStringLiteral("-")                              // output to stdout
    });

    proc.start(QIODevice::ReadOnly);

    if (!proc.waitForFinished(30000)) {
        qCWarning(lcAudioAnalyzer) << "ffmpeg timed out or failed to start for" << filePath;
        proc.kill();
        return result;
    }

    if (proc.exitCode() != 0) {
        qCWarning(lcAudioAnalyzer) << "ffmpeg exited with code" << proc.exitCode()
                                   << "for" << filePath;
        return result;
    }

    QByteArray pcmBuffer = proc.readAllStandardOutput();

    if (pcmBuffer.isEmpty()) {
        qCWarning(lcAudioAnalyzer) << "ffmpeg produced no audio data for" << filePath;
        return result;
    }

    // ── PCM data is now a contiguous float array at 8 kHz mono ─
    const auto *samples     = reinterpret_cast<const float *>(pcmBuffer.constData());
    const int   sampleCount = pcmBuffer.size() / static_cast<int>(sizeof(float));
    const int   sampleRate  = 8000;

    if (sampleCount == 0)
        return result;

    // Normalise
    float maxVal = 0.0f;
    for (int i = 0; i < sampleCount; ++i)
        maxVal = std::max(maxVal, std::abs(samples[i]));

    std::vector<float> normalised(sampleCount);
    if (maxVal > 0.0f) {
        for (int i = 0; i < sampleCount; ++i)
            normalised[i] = samples[i] / maxVal;
    } else {
        std::fill(normalised.begin(), normalised.end(), 0.0f);
    }

    // ── Chunk into ~chunkDuration windows and FFT each ─────────
    const int chunkSize = static_cast<int>(sampleRate * chunkDuration);
    const int numChunks = sampleCount / chunkSize;

    if (numChunks == 0)
        return result;

    const int fftSize = nextPow2(chunkSize);
    const auto window = hannWindow(chunkSize);

    // First pass: raw magnitudes per chunk per bar
    std::vector<std::vector<float>> rawFftData(numChunks);

    for (int c = 0; c < numChunks; ++c) {
        const int offset = c * chunkSize;

        // Zero-padded, windowed complex buffer
        std::vector<std::complex<float>> buf(fftSize, {0.0f, 0.0f});
        for (int i = 0; i < chunkSize; ++i)
            buf[i] = {normalised[offset + i] * window[i], 0.0f};

        fft_radix2(buf);

        // Magnitude of positive frequencies, skip DC
        // For a real signal FFT of size N, positive frequencies are bins 1..N/2-1
        // This matches Python's: fft = fft[1:len(fft)//2]  after rfft
        const int halfFFT = fftSize / 2;
        std::vector<float> mag(halfFFT - 1);
        for (int i = 0; i < static_cast<int>(mag.size()); ++i)
            mag[i] = std::abs(buf[i + 1]);   // +1 to skip DC

        if (mag.empty()) {
            rawFftData[c].assign(numBars, 0.0f);
            continue;
        }

        // Logarithmic bin edges — match Python's np.logspace(0, log10(len(fft)), numBars+1)
        // np.logspace(0, log10(N), M) produces M values from 10^0=1 to 10^log10(N)=N
        const int fftLen = static_cast<int>(mag.size());
        std::vector<int> logIndices(numBars + 1);
        for (int i = 0; i <= numBars; ++i) {
            double t = static_cast<double>(i) / numBars;
            double logVal = std::pow(static_cast<double>(fftLen), t);
            int idx = static_cast<int>(logVal);
            logIndices[i] = std::clamp(idx, 0, fftLen);
        }

        // Bin into bars
        std::vector<float> levels(numBars, 0.0f);
        for (int b = 0; b < numBars; ++b) {
            int startIdx = logIndices[b];
            int endIdx   = logIndices[b + 1];
            if (endIdx > startIdx) {
                float sum = 0.0f;
                for (int i = startIdx; i < endIdx; ++i)
                    sum += mag[i];
                levels[b] = sum / static_cast<float>(endIdx - startIdx);
            }
        }

        rawFftData[c] = std::move(levels);
    }

    // ── Global normalisation (95th percentile, matching Python) ─
    std::vector<float> allValues;
    allValues.reserve(numChunks * numBars);
    for (const auto &chunk : rawFftData)
        for (float v : chunk)
            if (v > 0.0f)
                allValues.push_back(v);

    float globalMax = 1.0f;
    if (!allValues.empty()) {
        std::sort(allValues.begin(), allValues.end());
        int idx95 = static_cast<int>(allValues.size() * 0.95);
        idx95 = std::min(idx95, static_cast<int>(allValues.size()) - 1);
        globalMax = allValues[idx95];
        if (globalMax <= 0.0f)
            globalMax = 1.0f;
    }

    // ── Second pass: normalise → 0-8 integer levels ────────────
    result.fftData.resize(numChunks);
    for (int c = 0; c < numChunks; ++c) {
        result.fftData[c].resize(numBars);
        for (int b = 0; b < numBars; ++b) {
            float norm = rawFftData[c][b] / globalMax;
            norm = std::min(1.0f, norm);
            norm = std::pow(norm, 0.6f);                   // perceptual curve
            result.fftData[c][b] = static_cast<int>(norm * 8.0f);
        }
    }

    result.success = true;
    return result;
}

// ════════════════════════════════════════════════════════════════
// Public slots
// ════════════════════════════════════════════════════════════════

void AudioAnalyzer::analyze_file(const QString &filePath)
{
    qCInfo(lcAudioAnalyzer) << "analyze_file called with:" << filePath;

    if (filePath.isEmpty() || !QFileInfo::exists(filePath)) {
        qCWarning(lcAudioAnalyzer) << "Invalid file path for analysis:" << filePath;
        return;
    }

    // Already analyzed this file
    if (filePath == m_currentFile && !m_fftData.empty()) {
        qCInfo(lcAudioAnalyzer) << "File already analyzed:" << filePath;
        return;
    }

    // Already busy
    if (m_analyzing) {
        qCInfo(lcAudioAnalyzer) << "Analysis already in progress, skipping";
        return;
    }

    m_currentFile = filePath;
    m_analyzing   = true;

    emit analysisStarted();
    qCInfo(lcAudioAnalyzer) << "Starting audio analysis for:" << filePath;

    // Capture locals for the lambda (numBars, chunkDuration)
    const int    bars     = m_numBars;
    const double chunkDur = m_chunkDuration;

    QFuture<AnalysisResult> future = QtConcurrent::run(
        [filePath, bars, chunkDur]() {
            return AudioAnalyzer::analyzeAudio(filePath, bars, chunkDur);
        });

    m_watcher.setFuture(future);
}

void AudioAnalyzer::onAnalysisDone()
{
    m_analyzing = false;

    AnalysisResult result = m_watcher.result();
    if (result.success) {
        m_fftData = std::move(result.fftData);
        emit analysisComplete();
        qCInfo(lcAudioAnalyzer) << "Analysis complete:"
                                << m_fftData.size() << "chunks generated";
    } else {
        m_fftData.clear();
        qCWarning(lcAudioAnalyzer) << "Analysis failed for" << m_currentFile;
    }
}

void AudioAnalyzer::update_position(float positionSeconds)
{
    if (m_fftData.empty() || !m_isActive)
        return;

    int chunkIndex = static_cast<int>(positionSeconds * 10.0f);
    chunkIndex = std::clamp(chunkIndex, 0, static_cast<int>(m_fftData.size()) - 1);

    const auto &newLevels = m_fftData[chunkIndex];

    // Build QVariantList and compare
    QVariantList varList;
    varList.reserve(static_cast<int>(newLevels.size()));
    for (int v : newLevels)
        varList.append(v);

    if (varList != m_currentLevels) {
        m_currentLevels = varList;
        emit fftDataChanged(m_currentLevels);
    }
}

void AudioAnalyzer::set_active(bool active)
{
    qCInfo(lcAudioAnalyzer) << "set_active called with:" << active
                            << ", has FFT data:" << m_fftData.size() << "chunks";
    m_isActive = active;
    if (!active) {
        // Emit zeros to clear the visualizer
        m_currentLevels.clear();
        m_currentLevels.reserve(m_numBars);
        for (int i = 0; i < m_numBars; ++i)
            m_currentLevels.append(0);
        emit fftDataChanged(m_currentLevels);
    }
}

QVariantList AudioAnalyzer::get_current_levels() const
{
    return m_currentLevels;
}

int AudioAnalyzer::get_num_bars() const
{
    return m_numBars;
}

bool AudioAnalyzer::is_analyzed() const
{
    return !m_fftData.empty();
}

void AudioAnalyzer::clear()
{
    m_fftData.clear();
    m_currentFile.clear();

    m_currentLevels.clear();
    m_currentLevels.reserve(m_numBars);
    for (int i = 0; i < m_numBars; ++i)
        m_currentLevels.append(0);
}

void AudioAnalyzer::set_quality(const QString &quality)
{
    int newBars = m_qualityBars.value(quality, 32);
    if (newBars == m_numBars)
        return;

    qCInfo(lcAudioAnalyzer) << "Visualizer quality changed to" << quality
                            << ":" << newBars << "bars";
    m_numBars = newBars;

    m_currentLevels.clear();
    m_currentLevels.reserve(m_numBars);
    for (int i = 0; i < m_numBars; ++i)
        m_currentLevels.append(0);

    // Clear cached analysis so next track uses new bar count
    m_fftData.clear();
    m_currentFile.clear();
}
